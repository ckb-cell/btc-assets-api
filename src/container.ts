import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
