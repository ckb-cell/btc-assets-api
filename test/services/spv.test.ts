import container from '../../src/container';
import { describe, test, beforeEach, afterEach, vi, expect } from 'vitest';
import BitcoinSPV from '../../src/services/spv';

describe('BitcoinSPV', () => {
  let bitcoinSPV: BitcoinSPV;

  beforeEach(async () => {
    const cradle = container.cradle;
    bitcoinSPV = new BitcoinSPV(cradle);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('getTxProof: throw BitcoinSPVError', async () => {
    vi.spyOn(bitcoinSPV['request'], 'post').mockResolvedValue({
      data: {
        jsonrpc: '2.0',
        error: {
          code: 23102,
          message:
            'target transaction is in header#2583777 and it requires 5 confirmations, but the tip header in local storage is header#2583781',
        },
        id: 'aa1a7882-9c0a-4eaa-87e8-1ed906b957f8',
      },
    });
    await expect(
      bitcoinSPV.getTxProof('ede749ecee5e607e761e4fffb6d754799498056872456a7d33abe426d7b9951c', 0, 100),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('getTxProof: get proof successfuly', async () => {
    vi.spyOn(bitcoinSPV['request'], 'post').mockResolvedValue({
      data: {
        jsonrpc: '2.0',
        result: {
          "spv_client": {
            "tx_hash": "0x5e570545c3ffd656199d3babd85f05377ac91b396126b166cf370e2f0edddae5",
            "index": "0x1"
          },
          "proof": "00000000000000"
        },
        id: 'aa1a7882-9c0a-4eaa-87e8-1ed906b957f8',
      },
    });
    const proof = await bitcoinSPV.getTxProof('ede749ecee5e607e761e4fffb6d754799498056872456a7d33abe426d7b9951c', 0, 100);
    expect(proof).toHaveProperty('spv_client');
    expect(proof).toHaveProperty('proof');
  });
});
