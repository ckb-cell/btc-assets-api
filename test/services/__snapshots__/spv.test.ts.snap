// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BitcoinSPV > getTxProof: throw BitcoinSPVError 1`] = `[BitcoinSPVError: target transaction is in header#2583777 and it requires 5 confirmations, but the tip header in local storage is header#2583781]`;
